src/prefect/_internal/concurrency/supervisors.py

```diff
@@ -144,16 +145,38 @@ class Supervisor(abc.ABC, Generic[T]):
     Work sent to the supervisor will be executed when the owner waits for the result.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, submit_fn: Callable[..., concurrent.futures.Future]) -> None:
+        self._submit_fn = submit_fn
         self.owner_thread_ident = threading.get_ident()
         self._future: AnyFuture = None
+        self._future_thread = concurrent.futures.Future()
         logger.debug("Created supervisor %r", self)
 
-    def set_future(self, future: AnyFuture) -> None:
+    def submit(self, __fn, *args, **kwargs) -> concurrent.futures.Future:
         """
-        Assign a future to the supervisor.
+        Submit a call to a worker.
+
+        The supervisor will use the `submit_fn` from initialization to send the call to
+        the worker. It will wrap the call to provide supervision.
+
+        This method may only be called once per supervisor.
         """
-        self._future = future
+        if self._future:
+            raise RuntimeError("A supervisor can only monitor a single future")
+
+        @functools.wraps(__fn)
+        def _call_in_supervised_thread():
+            # Capture the worker thread before running the function
+            thread = threading.current_thread()
+            self._future_thread.set_result(thread)
+            return __fn(*args, **kwargs)
```

Comment:
Here, we add tracking of the thread executing the future. We use a future to store this data so the supervisor can wait for it to be populated if it needs to.

Reply:
Cool trick :D